BuDDI software dev 0.2 pr

wli51 opened the pull request:
Added:
- more convenient pseudo-bulk generation (see updated example notebook 1 for details)
- buddi4 class for conveinent model initialization, saving/loading and storage of model architecture, primes for subsequent validation/analysis

nrosed reviewed (approved):
Amazing job -- all around I think it looks a lot cleaner and easier to use. Your example notebooks are also very easy to understand.

wli51 commented:
<p>The <code>get_supervised_dataset </code>and <code>get_unsupervised_dataset </code>methods are meant for bundling all the data modalities into tensorflow <code>dataset</code> objects to facilitate training. e.g. buddi4, once compiled, can be simply trained by </p><p> </p>

<pre class="ql-syntax" spellcheck="false">for _ in range(10):
  ds_sup_batch_input, ds_sup_batch_target = next(iter(ds_sup.batch(16).take(1))) # should be formal iteration
  ds_unsup_batch_input, ds_unsup_batch_target = next(iter(ds_unsup.batch(16).take(1))) # should be formal iteration
  
  compiled_buddi4_sup_model.train_on_batch((ds_sup_batch_input, ds_sup_batch_target))
  compiled_buddi4_sup_model.train_on_batch((ds_unsup_batch_input, ds_unsup_batch_target))

</pre><p> </p><p>The generated dataset objects can be shuffled or randomly down-sampled using native tensorflow functions </p>

<pre class="ql-syntax" spellcheck="false">shuffled = ds_sup_batch_input.shuffle(buffer_size=1000, seed=42)
downsampled = ds_sup_batch_input.take(200)
</pre><p> </p><p>Because of the nature of being batches and being of tensor datatype, outputs of <code>get_supervised_dataset </code>and <code>get_unsupervised_dataset </code>are not meant to be inputs of any plot/evaluation functions. The buddi_v2 plot functions take the raw X numpy array and pd.DataFrame metadata (and the user will need to manually filter these for data subsetting). I have not built any formal data structure class that allows for retrieval of data with sub-sampling (randomly but applied to all expression, proportion and metadata) or built-in filtering (e.g. return only expression, proportion and metadata corresponding to a specific perturbation) due to not carefully planning things out. </p><p> </p><p>I have a analysis specific class utility in the analysis repo (PR almost ready) that allows for separate access of single cell reference and bulk but nothing else. For all sorts of analysis/plotting there that may benefit from sub-sampling I just have the sub-samples built into the analysis helper functions (which is resulting in quite a bit of redundancy). Perhaps we can centralize the dataset manipulation, it shouldn't be a ton of work because I will just be wrapping numpy and pandas utility  </p>

---

 <a href='https://app.reviewnb.com/NDavidsonLab/buddi_v2/pull/2/discussion/#comment-2873269955'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

nrosed commented:
<p>aha -- I see. No worries, just wondering if it was possible with this function. If it doesn't take too much time to centralize the dataset manipulation, it may be worth it, but you can put it at a very low priority.</p>

---

 <a href='https://app.reviewnb.com/NDavidsonLab/buddi_v2/pull/2/discussion/#comment-2873269955'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

wli51 commented:
While I was waiting for my analysis stuff to run, I added some more notes/documentation here and there on what buddi_v2 datasets are for and are not for. I also modified the reparam layer to add a switch for seeded determinstic behavior that I find useful when putting together the analysis PR.

wli51 commented:
> another low-priority ask would also be to extend this to BuDDI-3. I am not sure how big of a lift this would be, so totally fine if you don't get around to it. It would be helpful though if you included some pointers on the steps to get BuDDI-3 working

I should be able to quickly pump-out a hard coded BuDDI-3 class and examples in 1-2 days (plan A). 



I have been working on a generalized BuDDI-x with any arbitrary number of latent spaces with the intension to include that in this PR (plan B). It ended up being very buggy so I had to gave up, but I think I have sorted out most of that (Also it would require much more a large infrastructure update for the dataset classes, preprocessing/augmentation).  

The realistic goal is probably scheduling plan A end of this week/next week.

wli51 commented:
Ended up adding the data container class to the PR. A lot of the down-stream analysis stuff would depend on it so if I were to come back to it later I would have to re-work all the down-stream analysis part after, which will be presumably more time consuming.

nrosed commented on src/buddi_v2/data/buddi4data.py:
```diff
@@ -0,0 +1,368 @@
+
+from ..plotting.plot_data import plot_data
+
+class BuDDI4Data:
```
looks great!

nrosed commented on src/buddi_v2/data/buddi4data.py:
```diff
@@ -0,0 +1,368 @@
+            self._selection[cond] = np.ones(n, dtype=bool)
+        return self
+
+    def query(self, conditions=None, **filters):
```
this is super useful
